KAFKA-14334: complete delayed purgatory after replication

jeffkbkim opened the pull request:
When a consumer makes a fetch request to a follower (for [KIP-392](https://cwiki.apache.org/confluence/display/KAFKA/KIP-392%3A+Allow+consumers+to+fetch+from+closest+replica#KIP392:Allowconsumerstofetchfromclosestreplica-ConsumerAPI)) the fetch request will sit in the purgatory until `fetch.max.wait.ms` because the purgatory is not completed after replication. This patch aims to complete the delayed fetch purgatory after successfully replicating from the leader.

Confirmed that the integration test fails when purgatory is not completed after replication.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

jeffkbkim committed:
complete delayed purgatory after replication

jeffkbkim committed:
fix createBrokerConfig compilation error

jeffkbkim committed:
remove unused method

jeffkbkim committed:
remove unused import

showuon commented on core/src/test/scala/integration/kafka/server/FetchFromFollowerIntegrationTest.scala:
```diff
@@ -0,0 +1,106 @@
+        }, "reassignment did not complete.")
+        admin.electLeaders(ElectionType.PREFERRED, Set(new TopicPartition(topic, 0)).asJava).all.get
+      }
+    })
```
I understand why you want to make broker1 as follower. But could we instead, set the `followerBrokerId` based on describeTopic result? So that we don't need partition reassignment and leader election. Ex:
```
val topicDescription = TestUtils.describeTopic(admin, topic)
topicDescription.partitions.forEach(partition => {
  leaderBrokerId = partition.leader().id()
  followerBrokerId = if (partition.leader().id() == 0) 1 else 0
})
```

jeffkbkim replied:
thanks for the suggestion. i followed david's suggestion to create a topic with replica placement

showuon commented on core/src/test/scala/integration/kafka/server/FetchFromFollowerIntegrationTest.scala:
```diff
@@ -0,0 +1,106 @@
+    // set fetch.max.wait.ms to a value greater than test utils wait time (15 seconds) to ensure that the
+    // delayed fetch purgatory is completed after successfully replicating from the leader.
+    val messages = initialMessages + nMessages
+    TestUtils.consumeTopicRecords(brokers, topic, messages, rackId = followerBrokerId.toString, fetchMaxWaitMs = 60000)
```
Could we reduce the `waitTime` here? Since currently, the test will take 15 seconds to fail, we can actually fail it after 5 seconds or less maybe?

jeffkbkim committed:
fix replica fetcher thread tests

showuon commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -331,6 +331,15 @@ class ReplicaManager(val config: KafkaConfig,
+   * from the leader for one or more partitions. Called in ReplicaFetcherThread after successfully
+   * replicating from the leader.
+   */
+  private[server] def completeDelayedFetchRequests(topicPartitions: Seq[TopicPartition]): Unit = {
```
maybe rename to : `maybeCompleteDelayedFetchRequests`, since in leader fetch case, the fetch purgatory won't be completed, right?

jeffkbkim replied:
are you referring to consumer fetch requests to the leader?

this method is only be called by replica fetchers, so no. however, the method itself does complete the fetch purgatory for the given topic partitions and there is no condition for when it doesn't. so i think a `maybe` is not appropriate.

showuon replied:
Ah, OK, yes, the `partitionsWithNewRecords` will be empty when consumer fetch from leader. Make sense.

showuon reviewed (commented):
@jeffkbkim , nice catch! Left some comments.

showuon commented on core/src/test/scala/integration/kafka/server/FetchFromFollowerIntegrationTest.scala:
```diff
@@ -0,0 +1,106 @@
+  @ParameterizedTest(name = TestInfoUtils.TestWithParameterizedQuorumName)
+  @ValueSource(strings = Array("zk", "kraft"))
+  def testFollowerCompleteDelayedPurgatoryOnReplication(quorum: String): Unit = {
+    TestUtils.generateAndProduceMessages(brokers, topic, nMessages)
+    // set fetch.max.wait.ms to a value greater than test utils wait time (15 seconds) to ensure that the
+    // delayed fetch purgatory is completed after successfully replicating from the leader.
+    val messages = initialMessages + nMessages
+    TestUtils.consumeTopicRecords(brokers, topic, messages, rackId = followerBrokerId.toString, fetchMaxWaitMs = 60000)
```
Actually, I don't think this test will fail reliably each time when something goes wrong. It relied on the consumer fetch will happen before follower fetch responded. Could we make it much more reliable?

jeffkbkim replied:
i'm not sure i follow. can you elaborate? if a consumer fetches before follower fetch responded, it should always sit in the purgatory and timeout, no?

instead of using a consumer, i updated the test to send a fetch request to the follower. it has a specified offset, so it will not be affected by the timing

dajac commented on core/src/test/scala/integration/kafka/server/FetchFromFollowerIntegrationTest.scala:
```diff
@@ -0,0 +1,106 @@
+  // Create a topic with a single partition. Access brokers via `brokers`
+  @BeforeEach
+  def initializeFetchFromFollowerCluster(): Unit = {
+    createTopic(topic, numParts, numNodes)
```
The best would be to create the topic with an assignment. This way, you can always say that broker 0 is the leader whereas the others are the followers. There is an example here: https://github.com/apache/kafka/blob/de0fb3c0fef3f83aafa1f3eb9fd403ed77bdf6dd/core/src/test/scala/unit/kafka/server/FetchRequestDownConversionConfigTest.scala#L117.

dajac commented on core/src/test/scala/integration/kafka/server/FetchFromFollowerIntegrationTest.scala:
```diff
@@ -0,0 +1,106 @@
+    // set fetch.max.wait.ms to a value greater than test utils wait time (15 seconds) to ensure that the
+    // delayed fetch purgatory is completed after successfully replicating from the leader.
+    val messages = initialMessages + nMessages
+    TestUtils.consumeTopicRecords(brokers, topic, messages, rackId = followerBrokerId.toString, fetchMaxWaitMs = 60000)
```
Did you consider sending a FetchRequest directly instead of relying on the consumer here? The downside of the using the consumer is that it does really guarantee that the fetch is serviced from the follower. It could be serviced by the leader if the cluster is misconfigured for instance.

jeffkbkim replied:
unrelated question: i was looking into verifying the follower's bytes out metric but was a bit confused. in https://github.com/apache/kafka/blob/trunk/core/src/test/scala/unit/kafka/metrics/MetricsTest.scala#L193 we set up a 2 broker cluster and access the bytes out metric using `TestUtils.meterCount()` which accesses the static yammer metrics registry. 

for integration tests, does this mean brokers' yammer metrics are merged together into one registry?

dajac commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -88,6 +92,11 @@ class ReplicaFetcherThread(name: String,
 
+  override def doWork(): Unit = {
+    super.doWork()
+    completeDelayedFetchRequests()
```
Could we add unit tests to cover this change?

dajac reviewed (commented):
@jeffkbkim Thanks for the patch. I left a few comments for consideration. I think that we should run a few benchmarks before merging this as well to ensure that this change does not have any bad side effects, especially when FFF is not enabled.

jeffkbkim committed:
create topic with replica placement, use fetch request

jeffkbkim commented on core/src/test/scala/integration/kafka/server/FetchFromFollowerIntegrationTest.scala:
```diff
@@ -0,0 +1,97 @@
+    val totalMessages = initialMessages + nMessages
+    val topicPartition = new TopicPartition(topic, 0)
+    val offsetMap = Map[TopicPartition, Long](
+      topicPartition -> (totalMessages - 1)
```
the test fails when the offset is set to `totalMessages`. i first assumed because it was due to delayed HWM propagation but looking at the response from `totalMessages - 1` shows
```
PartitionData(partitionIndex=0, errorCode=0, highWatermark=200, lastStableOffset=200, logStartOffset=0
```
so it does have HWM set to 200. why can't we fetch offset 200?

dajac replied:
I suppose that the request stays in the purgatory when you fetch offset 200 because there is not record at offset 200 yet and your don't produce anything after the fetch request is sent out.

I think that you test does not really test what you want here. In my mind, you should do the following:
* create the topic with no records.
* fetch at offset 0 with a high max wait. the request should be placed in the purgatory.
* produce one record.
* fetch request should be completed.

jeffkbkim replied:
the suggestion makes sense. i think the point you're getting at is we need to have the fetch sit in purgatory before replication. and i confirmed your suggestion works locally.

however, i'm still unable to grasp why fetching offset 200 fails. let's say "there is no record at offset 200 yet" which means replication is not complete. the fetch request will sit in the purgatory and wake up once the record is replicated.

could it be the case where record at offset 200 is already replicated but the follower's HWM is not updated yet to 200 so the fetch request sits in purgatory and never wakes up because there are no new records?

jeffkbkim replied:
ok, looking deeper into this, here's my understanding as to why fetching offset 200 did not work:

The log end offset points to the next offset the log will append a record, so technically it doesn't exist yet. furthermore, the high watermark is updated to the log end offset. in `Partition.maybeIncrementLeaderHW()`:
```
var newHighWatermark = leaderLogEndOffset
...
leaderLog.maybeIncrementHighWatermark(newHighWatermark)
```
even if the fetch offset is equal to the high watermark, we cannot fetch it if the high watermark is pointing to the log end offset. 

this is done in DelayedFetch where we skip accumulating bytes `if (endOffset.messageOffset (LEO / HWM) == fetchOffset.messageOffset (fetch offset))`

@dajac does that align with your understanding?

jeffkbkim replied:
also, i confirmed that the existing test fails when the follower is already up-to-date. the test above should work generally because the fetch request reaches the follower before the follower is fully replicated. will update the test

jeffkbkim commented:
thanks @showuon @dajac for the review. i have addressed the comments.

i am preparing a benchmark for the comparison.

dajac commented on core/src/test/scala/unit/kafka/server/ReplicaFetcherThreadTest.scala:
```diff
@@ -1100,6 +1100,66 @@ class ReplicaFetcherThreadTest {
+    appendInfo match {
+      case Some(_) =>
+        val argument: ArgumentCaptor[Seq[TopicPartition]] = ArgumentCaptor.forClass(classOf[Seq[TopicPartition]])
+        verify(replicaManager, times(1)).completeDelayedFetchRequests(argument.capture())
```
I am a bit surprised by this. Isn't it possible to do something like `verify(replicaManager, times(1)).completeDelayedFetchRequests(Seq(....))`?

dajac commented on core/src/test/scala/unit/kafka/server/ReplicaFetcherThreadTest.scala:
```diff
@@ -1100,6 +1100,66 @@ class ReplicaFetcherThreadTest {
+        verify(replicaManager, times(1)).completeDelayedFetchRequests(argument.capture())
+        assertEquals(Seq(tp0, tp1), argument.getValue)
+      case None =>
+        verify(replicaManager, times(0)).completeDelayedFetchRequests(any[Seq[TopicPartition]])
```
nit: Could we verify that the set is actually empty here?

jeffkbkim replied:
i'm not sure i follow. replicaManager.completeDelayedFetchRequests is not called when appendInfo is empty. i will add a check for both cases to assert that the underlying buffer is empty after thread.doWork

dajac replied:
ah, right. i missed the `times(0)`. my bad.

dajac commented on core/src/test/scala/unit/kafka/server/ReplicaFetcherThreadTest.scala:
```diff
@@ -1100,6 +1100,66 @@ class ReplicaFetcherThreadTest {
+    val tp1 = new TopicPartition("testTopic", 1)
+    val records = MemoryRecords.withRecords(CompressionType.NONE,
+      new SimpleRecord(1000, "foo".getBytes(StandardCharsets.UTF_8)))
+    val partitionData: thread.FetchData = new FetchResponseData.PartitionData()
+      .setRecords(records)
```
nit: In the case where `appendInfo` is `None`, it is a tad awkward that we set records here.

dajac commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -132,9 +141,17 @@ class ReplicaFetcherThread(name: String,
 
     brokerTopicStats.updateReplicationBytesIn(records.sizeInBytes)
 
+    logAppendInfo.foreach { _ => partitionsWithNewRecords += topicPartition }
```
I wonder if we could slightly better than this. If you look on the produce path, we have the following code:

```
  result.info.leaderHwChange match {
    case LeaderHwChange.Increased =>
      // some delayed operations may be unblocked after HW changed
      delayedProducePurgatory.checkAndComplete(requestKey)
      delayedFetchPurgatory.checkAndComplete(requestKey)
      delayedDeleteRecordsPurgatory.checkAndComplete(requestKey)
    case LeaderHwChange.Same =>
      // probably unblock some follower fetch requests since log end offset has been updated
      delayedFetchPurgatory.checkAndComplete(requestKey)
    case LeaderHwChange.None =>
      // nothing
  }
```

I wonder if we should do something similar here. Intuitively, we are interested by updated HWM and not so by new records, right? What do you think?

dajac replied:
Note that the HWM is updated separately here: `updateHighWatermarkAndStartOffset`.

dajac replied:
I am actually not sure about this. I need to think about it a little more. Let me know what you think.

jeffkbkim replied:
i think this makes sense. consumer fetches will at most fetch up to the HWM so we can add the topic partition only when `LeaderHwChange.Increased`

dajac replied:
Yeah, that's right. We just need to ensure that the HWM is always <= log end offset. Otherwise, this does not work. I will read this code again tomorrow.

jeffkbkim replied:
i believe that is the case. from [UnifiedLog.updatedHighWatermarkWithLogEndOffset()](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/log/UnifiedLog.scala#L597-L603)
```
  private def updateHighWatermarkWithLogEndOffset(): Unit = {
    // Update the high watermark in case it has gotten ahead of the log end offset following a truncation
    // or if a new segment has been rolled and the offset metadata needs to be updated.
    if (highWatermark >= localLog.logEndOffset) {
      updateHighWatermarkMetadata(localLog.logEndOffsetMetadata)
    }
  }
```
this is called on every `UnifiedLog.append()` and `UnifiedLog.roll()`

however, i did notice that `LeaderHWChange` is not set anywhere else except in the produce code path. from `Partition.appendRecordsToLeader()`:
```
info.copy(leaderHwChange = if (leaderHWIncremented) LeaderHwChange.Increased else LeaderHwChange.Same)
```
no other `LogAppendInfo` initialization passes in LeaderHWChange.

so, the follower would need to keep track of whether the high watermark changed for each partition which i don't think is something we want to do.

dajac replied:
I am still not fully convinced here. One aspect to remember is that the HWM is updated asynchronously and that the fetch request is bound by the HWM, not the log end offset. Let's imagine the following events:
1. The log is empty.
2. The consumer fetches from the follower at offset 0. The fetch request goes to the purgatory because the requested offset is not available yet.
3. The producer produces one request to the leader.
4. The follower fetches the new records and happens them to its local log. The log end offsets moves to offset 0 but the HWM stays at 0 for now. This trigger the purgatory because new records got appended.
5. The delayed fetch checks if the fetch request can be completed. It should not at this point because the HWM is still at offset 0.
6. The follower fetches again to commit the new records and to get the new HWM. It updates the HWM locally to offset 1. If there are no new records, we don't trigger the purgatory now. The delayed fetch remains in the purgatory until the timeout is reached.

This seems to be a legitimate course of actions, isn't it? Moreover, there is a race condition between 4 and 5 so the test may pass sometimes.

showuon replied:
Good point, David!

jeffkbkim replied:
> 4. The follower fetches the new records and happens them to its local log. The log end offsets moves to offset 0 but the HWM stays at 0 for now. This trigger the purgatory because new records got appended.

log end offset moves to offset 1 right?

i don't think there can be a race at step 5 & 6 (and not 4 & 5 right) because it's the same replica fetcher thread that updates the follower's HWM then completes the purgatory. that said, the current test has been working successfully consistently because
`logAppendInfo` in `ReplicaFetcherThread.processPartitionData()` 
```
    // Append the leader's messages to the log
    val logAppendInfo = partition.appendRecordsToFollowerOrFutureReplica(records, isFuture = false)
```
is not None. it's in the form of
```
LogAppendInfo(None,-1,None,-1,-1,-1,0,RecordConversionStats(temporaryMemoryBytes=0, numRecordsConverted=0, conversionTimeNanos=0),NoCompressionCodec,NoCompressionCodec,0,0,true,-1,List(),null,None)
```
so the follower was completing the purgatory both when it appended the new record, and again when it updated the HWM.

jeffkbkim replied:
@dajac i think in order to resolve the scenario you've mentioned, we would ideally only want to complete the purgatory when HWM is updated. and since HWM update is not asynchronous for the follower, it seems safe. what do you think?

dajac commented:
@jeffkbkim Thanks for the update. I left a few more comments/suggestions.

jeffkbkim committed:
send fetch request before replication in test

showuon commented on core/src/test/scala/unit/kafka/utils/TestUtils.scala:
```diff
@@ -818,7 +824,9 @@ object TestUtils extends Logging {
                            saslProperties: Option[Properties] = None,
                            keyDeserializer: Deserializer[K] = new ByteArrayDeserializer,
-                           valueDeserializer: Deserializer[V] = new ByteArrayDeserializer): KafkaConsumer[K, V] = {
+                           valueDeserializer: Deserializer[V] = new ByteArrayDeserializer,
+                           rackId: String = null,
+                           fetchMaxWaitMs: Int = 500): KafkaConsumer[K, V] = {
```
Is the test consumer change here and below necessary?

showuon commented on core/src/test/scala/integration/kafka/server/FetchFromFollowerIntegrationTest.scala:
```diff
@@ -0,0 +1,110 @@
+  @BeforeEach
+  def initializeFetchFromFollowerCluster(): Unit = {
+    // Create a 2 broker cluster where broker 0 is the leader and 1 is the follower.
+
```
nit: additional empty line

showuon commented on core/src/test/scala/integration/kafka/server/FetchFromFollowerIntegrationTest.scala:
```diff
@@ -0,0 +1,110 @@
+      topicPartition -> 0
+    )
+
+    val fetchRequest = createFetchRequest(
```
nit: Could we rename the variable to `consumerFetchRequest`? That would be much clear that we are sending consumer fetch, not follower fetch

showuon commented on core/src/test/scala/integration/kafka/server/FetchFromFollowerIntegrationTest.scala:
```diff
@@ -0,0 +1,110 @@
+
+  @ParameterizedTest(name = TestInfoUtils.TestWithParameterizedQuorumName)
+  @ValueSource(strings = Array("zk", "kraft"))
+  @Timeout(30)
```
Sorry, I think 30 seconds for a test to catch error is too long. Could we reduce to 15 secs? (remember to update the comment below, too)

jeffkbkim replied:
the test will actually fail after 45 seconds (when we get a response from the fetch request).

i've updated to 15 second timeout and 30 second fetch.max.wait.ms. should we reduce it further?

showuon reviewed (commented):
@jeffkbkim , thanks for the update. Left some comments.

dajac commented on core/src/test/scala/integration/kafka/server/FetchFromFollowerIntegrationTest.scala:
```diff
@@ -0,0 +1,110 @@
+import scala.jdk.CollectionConverters._
+
+class FetchFromFollowerIntegrationTest extends BaseFetchRequestTest {
+  val numNodes = 2
```
Could use 3 nodes here and 3 replicas for the partition? I think that 3 nodes will make the HWM propagation a bit slower so it could better highlight issues.

jeffkbkim replied:
as mentioned in the other thread, the purgatory is being completed on both replication and HWM update. have confirmed that the current setup always completes the purgatory once after appending new records and before HWM propagation because the replica fetcher thread does them sequentially.

jeffkbkim replied:
resolving this now that the follower completes the purgatory only when the HWM changes

dajac commented on core/src/test/scala/integration/kafka/server/FetchFromFollowerIntegrationTest.scala:
```diff
@@ -0,0 +1,110 @@
+  def initializeFetchFromFollowerCluster(): Unit = {
+    // Create a 2 broker cluster where broker 0 is the leader and 1 is the follower.
+
+    admin = TestUtils.createAdminClient(brokers, listenerName)
```
You could use `createAdminClient()` provided by `IntegrationTestHarness` here. The advantage is that the admin client returned is automatically closed. You don't have to worry about it.

jeffkbkim committed:
only complete purgatory on high watermark update

jeffkbkim commented on core/src/test/scala/unit/kafka/server/ReplicaFetcherThreadTest.scala:
```diff
@@ -1100,6 +1100,64 @@ class ReplicaFetcherThreadTest {
+    when(log.updateHighWatermark(newHighWatermark))
+      .thenReturn(HighWatermarkUpdate(newHighWatermark, hasHighWatermarkChanged))
+
+    val appendInfo: Option[LogAppendInfo] = Some(mock(classOf[LogAppendInfo]))
```
`val appendInfo = Partition.appendRecordstoFollowerOrFutureReplica()` can only be None when isFuture = true, i.e. in `ReplicaAlterLogDirsThread.processPartitionData()`. removed the test case for it

dajac commented on core/src/main/scala/kafka/log/UnifiedLog.scala:
```diff
@@ -379,7 +379,7 @@ class UnifiedLog(@volatile var logStartOffset: Long,
    * @return the updated high watermark offset
    */
-  def updateHighWatermark(hw: Long): Long = {
+  def updateHighWatermark(hw: Long): HighWatermarkUpdate = {
```
Did you consider adding a method like the following?

```
def maybeUpdateHighWatermark(hw: Long): Boolean = {
  lock.synchronized {
    val oldHighWatermark = fetchHighWatermarkMetadata
    val newHighWatermark = updateHighWatermark(LogOffsetMetadata(hw))
    oldHighWatermark.messageOffset != newHighWatermark.messageOffset
  }
}
```

This would be less intrusive. Is there any downside?

The nice thing is that you can directly do `if(log.maybeUpdateHighWatermark(...)) ...` on the caller side.

dajac commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -41,6 +43,9 @@ class ReplicaFetcherThread(name: String,
   this.logIdent = logPrefix
 
+  // Visible for testing
+  private[server] val partitionsWithNewRecords = mutable.Buffer[TopicPartition]()
```
nit: Should we rename this one? e.g. `partitionsWithNewHighWatermark`?

dajac commented on core/src/test/scala/integration/kafka/server/FetchFromFollowerIntegrationTest.scala:
```diff
@@ -0,0 +1,102 @@
+  def overridingProps: Properties = {
+    val props = new Properties
+    props.put(KafkaConfig.NumPartitionsProp, numParts.toString)
+
```
nit: We can remove this empty line.

dajac commented on core/src/test/scala/integration/kafka/server/FetchFromFollowerIntegrationTest.scala:
```diff
@@ -0,0 +1,102 @@
+    props
+  }
+
+  override def generateConfigs: collection.Seq[KafkaConfig] =
```
nit: Could we use curly braces here?

dajac commented on core/src/test/scala/integration/kafka/server/FetchFromFollowerIntegrationTest.scala:
```diff
@@ -0,0 +1,102 @@
+    val topicPartition = new TopicPartition(topic, 0)
+    val offsetMap = Map[TopicPartition, Long](
+      topicPartition -> 0
+    )
```
nit: You may be able to omit the type of the Map if you use `0L` for the offset.

dajac commented on core/src/test/scala/integration/kafka/server/FetchFromFollowerIntegrationTest.scala:
```diff
@@ -0,0 +1,102 @@
+      TestUtils.generateAndProduceMessages(brokers, topic, numMessages)
+      val response = receive[FetchResponse](socket, ApiKeys.FETCH, version)
+      assertEquals(Errors.NONE, response.error())
+      response.errorCounts().keySet().asScala.foreach { error => assertEquals(Errors.NONE, error)}
```
nit: This is usually not a good practice. It would be better to assert the entire collection: `assertEquals(Map(Errors.NONE -> 1), response.errorCounts()`.

dajac commented on core/src/test/scala/integration/kafka/server/FetchFromFollowerIntegrationTest.scala:
```diff
@@ -0,0 +1,102 @@
+      val response = receive[FetchResponse](socket, ApiKeys.FETCH, version)
+      assertEquals(Errors.NONE, response.error())
+      response.errorCounts().keySet().asScala.foreach { error => assertEquals(Errors.NONE, error)}
+
```
nit: Empty line could be removed here, I suppose.

dajac commented on core/src/test/scala/unit/kafka/server/BaseFetchRequestTest.scala:
```diff
@@ -46,10 +46,12 @@ class BaseFetchRequestTest extends BaseRequestTest {
   }
 
-  protected def createFetchRequest(maxResponseBytes: Int, maxPartitionBytes: Int, topicPartitions: Seq[TopicPartition],
+  protected def createConsumerFetchRequest(maxResponseBytes: Int, maxPartitionBytes: Int, topicPartitions: Seq[TopicPartition],
```
nit: Indentation of the arguments are off. While we are here, could you put on argument per line in order to be consistent?

```
protected def createConsumerFetchRequest(
  ...
  ...
  ...
): FetchRequest = { 
```

dajac commented on core/src/test/scala/unit/kafka/server/FetchRequestTest.scala:
```diff
@@ -51,7 +51,7 @@ class FetchRequestTest extends BaseFetchRequestTest {
 
     def createFetchRequest(topicPartitions: Seq[TopicPartition], offsetMap: Map[TopicPartition, Long] = Map.empty,
```
nit: Should we also rename this one?

dajac replied:
Thoughts?

jeffkbkim replied:
that makes sense. i updated all the other names as well

dajac commented on core/src/test/scala/unit/kafka/server/ReplicaFetcherThreadTest.scala:
```diff
@@ -1100,6 +1100,64 @@ class ReplicaFetcherThreadTest {
+  @Test
+  def testFetchResponseWithPartitionData(): Unit = {
+    verifyLocalFetchCompletionAfterHighWatermarkUpdate(hasHighWatermarkChanged = true)
+    verifyLocalFetchCompletionAfterHighWatermarkUpdate(hasHighWatermarkChanged = false)
```
Could we use?

```
  @ParameterizedTest
  @ValueSource(booleans = Array(true, false))
```

dajac commented on core/src/test/scala/integration/kafka/server/FetchFromFollowerIntegrationTest.scala:
```diff
@@ -0,0 +1,102 @@
+      topic,
+      brokers,
+      replicaAssignment = Map(0 -> Seq(leaderBrokerId, followerBrokerId))
+    )
```
Should we wait until the replicas are fully created with `TestUtils.waitForAllPartitionsMetadata` before proceeding?

jeffkbkim replied:
`TestUtils.createTopicWithAdmin()` invokes `waitForAllPartitionsMetadata()` inside

dajac commented on core/src/test/scala/unit/kafka/server/ReplicaFetcherThreadTest.scala:
```diff
@@ -1100,6 +1100,64 @@ class ReplicaFetcherThreadTest {
+    val tp1 = new TopicPartition("testTopic", 1)
+    val records = MemoryRecords.withRecords(CompressionType.NONE,
+      new SimpleRecord(1000, "foo".getBytes(StandardCharsets.UTF_8)))
+    val partitionData = new FetchResponseData.PartitionData().setRecords(records)
```
nit: We can probably call `setHighWatermark` here directly as it is common for both partitions.

```
val partitionData = new FetchResponseData.PartitionData()
  .setRecords(records)
  .setHighWatermark(newHighWatermark)
```

dajac commented on core/src/test/scala/unit/kafka/server/ReplicaFetcherThreadTest.scala:
```diff
@@ -1100,6 +1100,64 @@ class ReplicaFetcherThreadTest {
+  private def verifyLocalFetchCompletionAfterHighWatermarkUpdate(hasHighWatermarkChanged: Boolean): Unit = {
+    val props = TestUtils.createBrokerConfig(1, "localhost:1234")
+    val config = KafkaConfig.fromProps(props)
+    val newHighWatermark = 100L
```
nit: `newHighWatermark` sounds a bit wrong here because it is not new when `hasHighWatermarkChanged` is `false`. How about using `highWatermarkReceivedFromLeader` or something along those lines?

dajac commented on core/src/test/scala/integration/kafka/server/FetchFromFollowerIntegrationTest.scala:
```diff
@@ -0,0 +1,102 @@
+    val socket = connect(brokerSocketServer(followerBrokerId))
+    try {
+      send(fetchRequest, socket)
+      TestUtils.generateAndProduceMessages(brokers, topic, numMessages)
```
nit: How about inlining `numMessages = 1` here?

dajac commented on core/src/test/scala/integration/kafka/server/FetchFromFollowerIntegrationTest.scala:
```diff
@@ -0,0 +1,102 @@
+  @ParameterizedTest(name = TestInfoUtils.TestWithParameterizedQuorumName)
+  @ValueSource(strings = Array("zk", "kraft"))
+  @Timeout(15)
+  def testFollowerCompleteDelayedPurgatoryOnReplication(quorum: String): Unit = {
```
nit: `DelayedFetches` instead of `DelayedPurgatory`?

dajac commented on core/src/test/scala/integration/kafka/server/FetchFromFollowerIntegrationTest.scala:
```diff
@@ -0,0 +1,102 @@
+      .map(KafkaConfig.fromProps(_, overridingProps))
+
+  @BeforeEach
+  def initializeFetchFromFollowerCluster(): Unit = {
```
I wonder if we should just inline this topic creation knowing that we don't reuse it anywhere else. This has the advantage of keeping all the test logic in the test itself.

dajac reviewed (commented):
@jeffkbkim Thanks for the update. I left a few more comments.

artemlivshits commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -117,10 +127,11 @@ class ReplicaFetcherThread(name: String,
     // These values will be computed upon becoming leader or handling a preferred read replica fetch.
-    val followerHighWatermark = log.updateHighWatermark(partitionData.highWatermark)
+    val highWatermarkUpdate = log.updateHighWatermark(partitionData.highWatermark)
+    info(s"follower high watermark: ${highWatermarkUpdate.highWatermark}")
```
Should it be trace?  This could flood the info log. Also looks like we already log this info a few lines below.

artemlivshits commented on core/src/main/scala/kafka/log/UnifiedLog.scala:
```diff
@@ -379,7 +379,7 @@ class UnifiedLog(@volatile var logStartOffset: Long,
    * @param hw the suggested new value for the high watermark
    * @return the updated high watermark offset
```
We should update javadoc to doc the new return value.

jeffkbkim replied:
resolving as i've reverted the changes

artemlivshits commented on core/src/main/scala/kafka/log/UnifiedLog.scala:
```diff
@@ -390,7 +390,7 @@ class UnifiedLog(@volatile var logStartOffset: Long,
    * @param highWatermarkMetadata the suggested high watermark with offset metadata
    * @return the updated high watermark offset
```
Javadoc.

jeffkbkim replied:
ditto

jeffkbkim committed:
use maybeUpdateHighWatermark that returns whether hwm was updated

jeffkbkim commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -132,9 +139,22 @@ class ReplicaFetcherThread(name: String,
+    if (highWatermarkChanged) {
+      logAppendInfo.foreach { _ => partitionsWithNewHighWatermark += topicPartition }
+      if (logTrace)
+        trace(s"Follower updated replica high watermark for partition $topicPartition to ${partitionData.highWatermark}")
```
@dajac 

this statement is not entirely true and this was the reason i've attempted to use the HighWatermarkUpdated case class.  since this log is important, i think we should still have a case class for this.

dajac replied:
I see. How about returning an `Option` in `maybeUpdateHighWatermark` which would contain the updated HWM only if it was changed. Would this work?

I would also update the trace message to include both the received HWM from the leader and the updated HWM. Both seems useful here.

jeffkbkim replied:
much more elegant. thanks for the suggestion!

dajac commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -132,9 +139,22 @@ class ReplicaFetcherThread(name: String,
 
+    val highWatermarkChanged = log.maybeUpdateHighWatermark(partitionData.highWatermark)
+    if (highWatermarkChanged) {
+      logAppendInfo.foreach { _ => partitionsWithNewHighWatermark += topicPartition }
```
nit: I think that we don't have to do `logAppendInfo.foreach` here anymore, right?

dajac commented on core/src/test/scala/integration/kafka/server/FetchFromFollowerIntegrationTest.scala:
```diff
@@ -0,0 +1,95 @@
+  @ValueSource(strings = Array("zk", "kraft"))
+  @Timeout(15)
+  def testFollowerCompleteDelayedFetchesOnReplication(quorum: String): Unit = {
+    // Create a 2 broker cluster where broker 0 is the leader and 1 is the follower.
```
nit: Should we say `Create a topic with 2 replicas where...`? The cluster is automatically created.

dajac commented on core/src/test/scala/integration/kafka/server/FetchFromFollowerIntegrationTest.scala:
```diff
@@ -0,0 +1,95 @@
+      replicaAssignment = Map(0 -> Seq(leaderBrokerId, followerBrokerId))
+    )
+
+    // Set fetch.max.wait.ms to a value (20 seconds) greater than the timeout (15 seconds).
+    // Send a fetch request before the record is replicated to ensure that the replication
+    // triggers purgatory completion.
```
nit: Should we move this to right before `val fetchRequest` as the comment is for this part?

dajac commented on core/src/main/scala/kafka/log/UnifiedLog.scala:
```diff
@@ -433,6 +433,22 @@ class UnifiedLog(@volatile var logStartOffset: Long,
     }
   }
 
+  /**
```
nit: Could we also update the scaladoc of `def updateHighWatermark(hw: Long)` to remove the part about the follower updating the HWM. This is no longer the case.

dajac reviewed (commented):
Thanks for the update. I left a few more minor comments.

jeffkbkim committed:
return Option[Long] maybeUpdateHighWatermark(), address comments

jeffkbkim commented:
@dajac thanks for the review. i have addressed the comments.

showuon commented on core/src/test/scala/unit/kafka/server/ReplicaFetcherThreadTest.scala:
```diff
@@ -1100,6 +1102,66 @@ class ReplicaFetcherThreadTest {
+    thread.doWork()
+
+    if (highWatermarkUpdated) {
+      verify(replicaManager, times(1)).completeDelayedFetchRequests(Seq(tp0, tp1))
```
nit: Could we verify `completeDelayedFetchRequets` is not called before invoking `doWork` ?

showuon reviewed (approved):
@jeffkbkim , thanks for the update. LGTM! Only a minor comment.

dajac commented on core/src/main/scala/kafka/log/UnifiedLog.scala:
```diff
@@ -433,6 +432,27 @@ class UnifiedLog(@volatile var logStartOffset: Long,
+   *
+   * @return the new high watermark if the high watermark changed, None otherwise.
+   */
+  def maybeUpdateHighWatermark(hw: Long): Option[Long] = {
```
Should we add unit tests in `UnifiedLogTest` for this new method?

dajac commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -132,9 +139,22 @@ class ReplicaFetcherThread(name: String,
 
     brokerTopicStats.updateReplicationBytesIn(records.sizeInBytes)
 
+    log.maybeUpdateHighWatermark(partitionData.highWatermark).foreach { newHighWatermark =>
```
Note that we are changing the order here as `updateHighWatermark` was called before `maybeIncrementLogStartOffset`. It may be better to keep the original order in order to avoid unexpected side effects. It probably does not matter but I rather prefer to be safe.

dajac commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -132,9 +139,22 @@ class ReplicaFetcherThread(name: String,
+    log.maybeUpdateHighWatermark(partitionData.highWatermark).foreach { newHighWatermark =>
+      partitionsWithNewHighWatermark += topicPartition
+      if (logTrace)
+        trace(s"Follower received high watermark ${partitionData.highWatermark} from the leader and " +
+          s"updated replica high watermark to $newHighWatermark for partition $topicPartition")
```
Should we also log something when the HWM is not changed? Tracing the HWM received from the leader could be useful in some situations. We could perhaps say something like: `Follower received high watermark {} from the leader but did not update replica high watermark for partition {}`.

dajac reviewed (commented):
@jeffkbkim Thanks for your patience on this one. Here is a last round of comments. We should be good to go afterwards.

jeffkbkim committed:
address comments

dajac commented on core/src/test/scala/integration/kafka/server/FetchFromFollowerIntegrationTest.scala:
```diff
@@ -0,0 +1,94 @@
+      send(fetchRequest, socket)
+      TestUtils.generateAndProduceMessages(brokers, topic, numMessages = 1)
+      val response = receive[FetchResponse](socket, ApiKeys.FETCH, version)
+      assertEquals(Errors.NONE, response.error())
```
nit: We could omit the `()` after `error`.

dajac commented on core/src/test/scala/unit/kafka/log/UnifiedLogTest.scala:
```diff
@@ -3455,6 +3455,21 @@ class UnifiedLogTest {
+      log.appendAsLeader(records, leaderEpoch = 0)
+    }
+
+    assertEquals(Some(100L), log.maybeUpdateHighWatermark(100L))
```
nit: Could we add `assertEquals(Some(99L), log.maybeUpdateHighWatermark(99L))` as well?

dajac replied:
The sequence is a bit weird. I would have added `99L` just before `100L`. Something like this:

```
    assertEquals(Some(99L), log.maybeUpdateHighWatermark(99L))
    assertEquals(None, log.maybeUpdateHighWatermark(99L))

    assertEquals(Some(100L), log.maybeUpdateHighWatermark(100L))
    assertEquals(None, log.maybeUpdateHighWatermark(100L))

    // Bound by the log end offset.
    assertEquals(None, log.maybeUpdateHighWatermark(101L))
```

dajac commented on core/src/test/scala/integration/kafka/server/FetchFromFollowerIntegrationTest.scala:
```diff
@@ -0,0 +1,94 @@
+  val topic = "test-fetch-from-follower"
+  val leaderBrokerId = 0
+  val followerBrokerId = 1
+  var admin: Admin = null
```
nit: `admin` is no longer used outside of the test. We could use a local var now.

dajac commented on core/src/test/scala/unit/kafka/server/ReplicaFetcherThreadTest.scala:
```diff
@@ -39,12 +39,14 @@ import org.junit.jupiter.api.{AfterEach, Test}
 import java.util
 import java.util.{Collections, Optional}
-
 import org.apache.kafka.server.common.MetadataVersion
 import org.apache.kafka.server.common.MetadataVersion.IBP_2_6_IV0
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.ValueSource
```
While we are here, could we move these imports up? We already have `org.apache.kafka` and `org.junit.jupiter` imports above.

dajac reviewed (commented):
@jeffkbkim Thanks for the update. The patch LGTM. I left a few nits and it seems that a few unit test failed before `kafka.log.UnifiedLog.maybeUpdateHighWatermark(long)` is not mocked. Could you fix those?

jeffkbkim committed:
fix tests, tidy

jeffkbkim committed:
fix compilation error

jeffkbkim committed:
fix test failure

jeffkbkim commented:
@dajac thanks for the review, i have addressed the comments. i'm a bit confused since the new parameterized replica fetcher test seems to be failing in jdk 8 / scala 2.12

dajac commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -135,6 +151,13 @@ class ReplicaFetcherThread(name: String,
 
+  private def completeDelayedFetchRequests(): Unit = {
+    if (partitionsWithNewHighWatermark.nonEmpty) {
+      replicaMgr.completeDelayedFetchRequests(partitionsWithNewHighWatermark.toSeq)
```
I think that the failed test with Java 1.8 and Scala 2.12 is due to this line. In Scala 2.13, `toSeq` creates an immutable copy where it does not in 2.12. Hence in 2.12, the collection is cleared at the next line before the mock is verified. We should force the copy here (and add a comment).

jeffkbkim replied:
i'm concerned that copying the buffer will affect performance. Should we use `toIndexedSeq`?

dajac replied:
Yeah, I agree with you. Perhaps, we could actually change the test to cope with this instead of changing the main code. A potential solution is discussed here: https://stackoverflow.com/questions/50025561/matching-mutable-object-without-argumentcaptor. What do you think?

jeffkbkim replied:
that makes sense, thanks for the suggestion!

dajac commented on core/src/test/scala/unit/kafka/server/ReplicaFetcherThreadTest.scala:
```diff
@@ -1100,6 +1102,67 @@ class ReplicaFetcherThreadTest {
+    val mockBlockingSend: BlockingSend = mock(classOf[BlockingSend])
+    when(mockBlockingSend.brokerEndPoint()).thenReturn(brokerEndPoint)
+
+    var maybeNewHighWatermark: Option[Long] = None
+    if (highWatermarkUpdated) {
+      maybeNewHighWatermark = Some(highWatermarkReceivedFromLeader)
+    }
```
small nit: In Scala, we tend to write this as follow:

```
val maybeNewHighWatermark = if (highWatermarkUpdated) {
  Some(highWatermarkReceivedFromLeader)
} else {
  None
}
```

It may also fit on one line.

jeffkbkim committed:
address comments, fix 2.12 test failure

jeffkbkim committed:
fix scala 2.12 failure

dajac commented:
@jeffkbkim There are failed tests. Could you take a look please?

jeffkbkim commented:
new test failures
```
ConsumerCoordinatorTest.testPrepareJoinAndRejoinAfterFailedRebalance
ProducerIdExpirationTest.testTransactionAfterTransactionIdExpiresButProducerIdRemains
```
and existing test failure
```
PlaintextAdminIntegrationTest.testAlterReplicaLogDirs
```
pass locally. the failures do not seem related to this PR

dajac commented:
@jeffkbkim `testAlterReplicaLogDirs(String)` has failed in the last few builds. I don't see the same in other PRs. Are you sure that it is not related? I haven't had the time to investigate it myself.

jeffkbkim commented:
@dajac my bad, i was looking at the wrong test. it's related

jeffkbkim commented:
error from `testAlterReplicaLogDirs(String)`:
```
[2022-11-15 10:24:37,184] ERROR [ReplicaFetcher replicaId=2, leaderId=0, fetcherId=0] Unexpected error occurred while processing data for partition topic-0 at offset 15 (kafka.server.ReplicaFetcherThread:76)
java.lang.NullPointerException
	at kafka.log.OffsetIndex.$anonfun$lookup$1(OffsetIndex.scala:90)
...
```
we hit NPE because the replica fetcher thread tries to find the high watermark from the to-be-deleted log segment's offset index, where the currently replicating segment's offset index mmap is null. the offset index is unmapped during `LogManager.replaceCurrentWithFutureLog(TopicPartition)`.

in UnifiedLog, we check whether the memory map buffer is closed when retrieiving HWM:
```
  private def fetchHighWatermarkMetadata: LogOffsetMetadata = {
    localLog.checkIfMemoryMappedBufferClosed()
    ...
  }
```
which makes it seem okay to read from the offset index. however, we should not allow any disk io if an offset index is unmapped. moreover, if a local log is closed, we should set `isMemoryMappedBufferClosed` to `true` in `LocalLog.close()`:
```
  private[log] def close(): Unit = {
    maybeHandleIOException(s"Error while renaming dir for $topicPartition in dir ${dir.getParent}") {
      checkIfMemoryMappedBufferClosed()
      segments.close()
      // isMemoryMappedBufferClosed = true?
    }
  }
```

@dajac i'm not sure if this is a bug or there's another reason we do not set the mmap flag when closing a log. is my understanding correct? if so, i can create a jira. besides altering replica log dir UnifiedLog.close() is only called during shutdown.

as for the replica fetcher, maybe we can directly retrieve `highWatermarkMetadata` like
```
  def maybeUpdateHighWatermark(hw: Long): Option[Long] = {
    lock.synchronized {
      val oldHighWatermark = highWatermarkMetadata
      updateHighWatermark(LogOffsetMetadata(hw)) match {
        case oldHighWatermark.messageOffset =>
          None
        case newHighWatermark =>
          Some(newHighWatermark)
      }
    }
  }
``` 
instead of calling `fetchHighWatermarkMetadata()` for the old HWM since the follower does not care about the HWM's base offset or relative position in segment.

jeffkbkim committed:
use highWatermarkMetadata

dajac reviewed (approved):
LGTM, thanks @jeffkbkim!

dajac commented:
Failed tests are unrelated:
```
Build / JDK 17 and Scala 2.13 / kafka.api.TransactionsExpirationTest.testTransactionAfterProducerIdExpires(String).quorum=kraft
Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest.testCreatePartitionsUseProvidedForwardingAdmin()
Build / JDK 11 and Scala 2.13 / org.apache.kafka.controller.QuorumControllerTest.testDelayedConfigurationOperations()
Build / JDK 11 and Scala 2.13 / org.apache.kafka.trogdor.coordinator.CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated()
Build / JDK 8 and Scala 2.12 / org.apache.kafka.controller.QuorumControllerTest.testEarlyControllerResults()
```

dajac commented:
Merged to trunk and 3.3.

ijuma commented on core/src/test/scala/unit/kafka/utils/TestUtils.scala:
```diff
@@ -214,14 +214,14 @@ object TestUtils extends Logging {
-    startingIdNumber: Int = 0
-  ): Seq[Properties] = {
+    startingIdNumber: Int = 0,
+    enableFetchFromFollower: Boolean = false): Seq[Properties] = {
```
This method has far too many parameters, it doesn't make sense to add parameters for every single possible configuration. Instead, you can add whatever configuration you want to the returned `Properties` instance.
